File: catalog/query_set.py
"""Strict application-level QuerySet contract for LLM query modification."""

from copy import deepcopy
import math


class QuerySetValidationError(ValueError):
    pass


def _range():
    return {"min": None, "max": None}


QUERY_SET_TEMPLATE = {
    "brand": None,
    "model": None,
    "release_date": None,
    "source": {"name": None, "url": None},
    "performance": {
        "chipset": None, "cpu": None, "gpu": None, "storage_type": None,
        "variants": {"ram_gb": _range(), "storage_gb": _range()},
    },
    "display": {
        "size_inches": _range(), "resolution_width": _range(), "resolution_height": _range(),
        "technology": None, "refresh_rate_hz": _range(), "brightness_peak_nits": _range(), "hdr": None,
    },
    "battery": {"capacity_mah": _range(), "charging_w": _range(), "wireless_charging": None},
    "camera": {
        "main_mp": _range(), "ultrawide_mp": _range(), "macro_mp": _range(), "selfie_mp": _range(),
        "ois": None, "video_max_resolution": None, "video_max_fps": _range(),
    },
    "connectivity": {"5g": None, "wifi_version": None, "bluetooth_version": None, "nfc": None},
    "physical": {"weight_g": _range(), "ip_rating": None},
    "software": {"os": None, "android_version": _range(), "major_updates": _range()},
    "benchmarks": {"antutu": _range(), "geekbench": _range(), "3dmark": _range()},
    "price": _range(),
}

BOOLEAN_PATHS = {
    ("display", "hdr"), ("battery", "wireless_charging"), ("camera", "ois"),
    ("connectivity", "5g"), ("connectivity", "nfc"),
}


def empty_query_set():
    return deepcopy(QUERY_SET_TEMPLATE)


def _validate(value, template, path=()):
    label = ".".join(path) or "root"
    if isinstance(template, dict):
        if not isinstance(value, dict):
            raise QuerySetValidationError(f"{label} must be an object")
        expected, actual = set(template), set(value)
        if expected != actual:
            unknown = sorted(actual - expected)
            missing = sorted(expected - actual)
            detail = []
            if unknown:
                detail.append(f"unknown fields: {', '.join(unknown)}")
            if missing:
                detail.append(f"missing fields: {', '.join(missing)}")
            raise QuerySetValidationError(f"{label} has invalid structure ({'; '.join(detail)})")
        for key in template:
            _validate(value[key], template[key], path + (key,))
        if set(template) == {"min", "max"}:
            minimum, maximum = value["min"], value["max"]
            if minimum is not None and maximum is not None and minimum > maximum:
                raise QuerySetValidationError(f"{label}.min cannot exceed {label}.max")
        return
    if path[-1] in {"min", "max"} and value is not None:
        if (
            isinstance(value, bool)
            or not isinstance(value, (int, float))
            or not math.isfinite(value)
            or value < 0
        ):
            raise QuerySetValidationError(f"{label} must be a non-negative finite number or null")
    elif path in BOOLEAN_PATHS:
        if value is not None and not isinstance(value, bool):
            raise QuerySetValidationError(f"{label} must be boolean or null")
    elif template is None and value is not None and not isinstance(value, str):
        raise QuerySetValidationError(f"{label} must be text or null")


def validate_query_set(query_set):
    """Validate and deep-copy a value before it crosses into the domain layer."""
    _validate(query_set, QUERY_SET_TEMPLATE)
    return deepcopy(query_set)

File: frontend/src/features/torobche/query-set.ts
import type { NumericRange, TorobcheQuerySet } from "@/types/api";

export interface ActiveCriterion {
  path: string;
  label: string;
  value: string;
}

const labels: Record<string, string> = {
  brand: "برند",
  model: "مدل",
  release_date: "تاریخ عرضه",
  "performance.chipset": "تراشه",
  "performance.cpu": "پردازنده",
  "performance.gpu": "پردازندهٔ گرافیکی",
  "performance.storage_type": "نوع حافظه",
  "performance.variants.ram_gb": "رم",
  "performance.variants.storage_gb": "حافظه",
  "display.size_inches": "اندازهٔ نمایشگر",
  "display.technology": "فناوری نمایشگر",
  "display.refresh_rate_hz": "نرخ نوسازی",
  "display.brightness_peak_nits": "روشنایی",
  "display.hdr": "HDR",
  "battery.capacity_mah": "باتری",
  "battery.charging_w": "شارژ",
  "battery.wireless_charging": "شارژ بی‌سیم",
  "camera.main_mp": "دوربین اصلی",
  "camera.selfie_mp": "دوربین سلفی",
  "camera.ois": "لرزش‌گیر اپتیکال",
  "connectivity.5g": "شبکهٔ 5G",
  "connectivity.nfc": "NFC",
  "physical.weight_g": "وزن",
  "physical.ip_rating": "مقاومت",
  "software.os": "سیستم‌عامل",
  price: "بازهٔ قیمت",
};

const units: Record<string, string> = {
  "performance.variants.ram_gb": "GB",
  "performance.variants.storage_gb": "GB",
  "display.size_inches": "اینچ",
  "display.refresh_rate_hz": "Hz",
  "display.brightness_peak_nits": "nit",
  "battery.capacity_mah": "mAh",
  "battery.charging_w": "W",
  "camera.main_mp": "MP",
  "camera.selfie_mp": "MP",
  "physical.weight_g": "g",
  price: "تومان",
};

function isRange(value: unknown): value is NumericRange {
  return typeof value === "object" && value !== null && "min" in value && "max" in value;
}

function formatValue(path: string, value: string | boolean | NumericRange): string | null {
  if (typeof value === "boolean") return value ? "دارد" : "ندارد";
  if (typeof value === "string") return value;
  if (value.min === null && value.max === null) return null;
  const unit = units[path] ? ` ${units[path]}` : "";
  if (value.min !== null && value.max !== null)
    return `${value.min.toLocaleString("fa-IR")} تا ${value.max.toLocaleString("fa-IR")}${unit}`;
  if (value.min !== null) return `از ${value.min.toLocaleString("fa-IR")}${unit}`;
  return `تا ${value.max?.toLocaleString("fa-IR")}${unit}`;
}

export function activeCriteria(querySet: TorobcheQuerySet): ActiveCriterion[] {
  const criteria: ActiveCriterion[] = [];
  function walk(value: unknown, path: string) {
    if (value === null || path.startsWith("source")) return;
    if (typeof value === "string" || typeof value === "boolean" || isRange(value)) {
      const formatted = formatValue(path, value);
      if (formatted)
        criteria.push({
          path,
          label: labels[path] ?? path.split(".").at(-1) ?? path,
          value: formatted,
        });
      return;
    }
    if (typeof value === "object")
      Object.entries(value).forEach(([key, child]) => walk(child, path ? `${path}.${key}` : key));
  }
  walk(querySet, "");
  return criteria;
}
